diff -urB cominisatps-SAT2016/core/Solver.cc cominisatps/core/Solver.cc
--- cominisatps-SAT2016/core/Solver.cc	2016-05-02 03:22:00.000000000 +0200
+++ cominisatps/core/Solver.cc	2020-07-17 09:34:07.994400936 +0200
@@ -150,6 +150,7 @@
     trail    .capacity(v+1);
     setDecisionVar(v, dvar);
 
+    assump.push(false);
     // Additional space needed for stamping.
     // TODO: allocate exact memory.
     seen      .push(0);
@@ -550,16 +551,20 @@
 |    Specialized analysis procedure to express the final conflict in terms of assumptions.
 |    Calculates the (possibly empty) set of assumptions that led to the assignment of 'p', and
 |    stores the result in 'out_conflict'.
+|
+|  void Solver::analyzeFinal(Lit p, vec<Lit>& out_conflict)
 |________________________________________________________________________________________________@*/
-void Solver::analyzeFinal(Lit p, vec<Lit>& out_conflict)
+void Solver::analyzeFinal(CRef confl, vec<Lit>& out_conflict)
 {
     out_conflict.clear();
-    out_conflict.push(p);
 
     if (decisionLevel() == 0)
         return;
 
-    seen[var(p)] = 1;
+    Clause& cl = ca[confl];
+    for (int j = 0; j < cl.size(); j++)
+        if (level(var(cl[j])) > 0)
+            seen[var(cl[j])] = 1;
 
     for (int i = trail.size()-1; i >= trail_lim[0]; i--){
         Var x = var(trail[i]);
@@ -577,7 +582,9 @@
         }
     }
 
-    seen[var(p)] = 0;
+    for (int j = cl.size() == 2 ? 0 : 1; j < cl.size(); j++)
+        if (level(var(cl[j])) > 0)
+            seen[var(cl[j])] = 0;
 }
 
 
@@ -650,12 +657,22 @@
                 *j++ = w; continue; }
 
             // Look for new watch:
+            int choosenPos = -1, maxAssumpLevel = assumptions.size() > 0 ? 1 : 0;
             for (int k = 2; k < c.size(); k++)
                 if (value(c[k]) != l_False){
-                    c[1] = c[k]; c[k] = false_lit;
-                    watches[~c[1]].push(w);
-                    goto NextClause; }
-
+                    choosenPos = k;
+                    if(decisionLevel()>maxAssumpLevel || value(c[k])==l_True || !assump[var(c[k])]) 
+                        break; }
+            if(choosenPos!=-1) {
+                c[1] = c[choosenPos]; c[choosenPos] = false_lit;
+                watches[~c[1]].push(w);
+                goto NextClause; }
+             /*   for (int k = 2; k < c.size(); k++)
+                    if (value(c[k]) != l_False){
+                        c[1] = c[k]; c[k] = false_lit;
+                        watches[~c[1]].push(w);
+                        goto NextClause; }
+*/
             // Did not find watch -- clause is unit under assignment:
             *j++ = w;
             if (value(first) == l_False){
@@ -789,7 +806,8 @@
         ok = stampAll(true);
 
     checkGarbage();
-    rebuildOrderHeap();
+    if ((int)dec_vars - nAssigns() < (glucose_restart ? order_heap_glue_r : order_heap_no_r).size() / 2) // M. Piotrow 20.07.2017
+        rebuildOrderHeap();
 
     simpDB_assigns = nAssigns();
     simpDB_props   = clauses_literals + learnts_literals;   // (shouldn't depend on stats really, but it will do for now)
@@ -933,6 +951,7 @@
             conflicts++; nof_conflicts--;
             if (conflicts == 100000 && learnts_core.size() < 100) core_lbd_cut = 5;
             if (decisionLevel() == 0) return l_False;
+            if (decisionLevel() == 1) { analyzeFinal(confl, conflict); return l_False; }
 
             learnt_clause.clear();
             analyze(confl, learnt_clause, backtrack_level, lbd);
@@ -998,7 +1017,7 @@
                 restart = lbd_queue.full() && (lbd_queue.avg() * K > global_lbd_sum / conflicts_glue);
                 cached = true;
             }
-            if (restart /*|| !withinBudget()*/){
+            if (restart || !withinBudget()){
                 lbd_queue.clear();
                 // Reached bound on number of conflicts:
                 progress_estimate = progressEstimate();
@@ -1017,22 +1036,18 @@
                 reduceDB(); }
 
             Lit next = lit_Undef;
-            /*while (decisionLevel() < assumptions.size()){
-                // Perform user provided assumption:
-                Lit p = assumptions[decisionLevel()];
-                if (value(p) == l_True){
-                    // Dummy decision level:
-                    newDecisionLevel();
-                }else if (value(p) == l_False){
-                    analyzeFinal(~p, conflict);
-                    return l_False;
-                }else{
-                    next = p;
-                    break;
+
+            if (decisionLevel() == 0) {
+                newDecisionLevel();
+                for (int i = 0; i < assumptions.size(); i++) {
+                    Lit p = assumptions[i];
+                    if (value(p) == l_False) { conflict.push(~p); return l_False; }
+                    else if (value(p) != l_True) uncheckedEnqueue(p);
                 }
+                continue;
             }
 
-            if (next == lit_Undef)*/{
+            if (next == lit_Undef){
                 // New variable decision:
                 decisions++;
                 next = pickBranchLit();
@@ -1101,25 +1116,23 @@
 
     solves++;
 
+    for (int i = 0; i <assumptions.size(); i++)
+        assump[var(assumptions[i])] = true;
+
     max_learnts               = nClauses() * learntsize_factor;
     learntsize_adjust_confl   = learntsize_adjust_start_confl;
     learntsize_adjust_cnt     = (int)learntsize_adjust_confl;
     lbool   status            = l_Undef;
 
-    if (verbosity >= 1){
-        printf("c ============================[ Search Statistics ]==============================\n");
-        printf("c | Conflicts |          ORIGINAL         |          LEARNT          | Progress |\n");
-        printf("c |           |    Vars  Clauses Literals |    Limit  Clauses Lit/Cl |          |\n");
-        printf("c ===============================================================================\n");
-    }
+    if (verbosity >= 1)
+        printf("c ======================[ COMiniSatPS search starting  ]========================\n");
 
     add_tmp.clear();
-
     glucose_restart = true;
     int init = 10000;
-    while (status == l_Undef && init > 0 /*&& withinBudget()*/)
+    while (status == l_Undef && init > 0 && withinBudget())
        status = search(init);
-    glucose_restart = false;
+    if (status == l_Undef && withinBudget()) glucose_restart = false;
 
     // Search:
     int phase_allotment = 100;
@@ -1127,9 +1140,9 @@
         int weighted = glucose_restart ? phase_allotment * 2 : phase_allotment;
         fflush(stdout);
 
-        while (status == l_Undef && weighted > 0 /*&& withinBudget()*/)
+        while (status == l_Undef && weighted > 0 && withinBudget())
             status = search(weighted);
-        if (status != l_Undef /*|| !withinBudget()*/)
+        if (status != l_Undef || !withinBudget())
             break; // Should break here for correctness in incremental SAT solving.
 
         glucose_restart = !glucose_restart;
@@ -1137,9 +1150,6 @@
             phase_allotment += phase_allotment / 10;
     }
 
-    if (verbosity >= 1)
-        printf("c ===============================================================================\n");
-
 #ifdef BIN_DRUP
     if (drup_file && status == l_False) binDRUP_flush(drup_file);
 #endif
diff -urB cominisatps-SAT2016/core/Solver.h cominisatps/core/Solver.h
--- cominisatps-SAT2016/core/Solver.h	2016-05-02 03:22:00.000000000 +0200
+++ cominisatps/core/Solver.h	2020-07-17 09:34:07.994400936 +0200
@@ -23,7 +23,7 @@
 #ifndef Minisat_Solver_h
 #define Minisat_Solver_h
 
-#define BIN_DRUP
+//#define BIN_DRUP
 
 #define GLUCOSE23
 //#define EXTRA_VAR_ACT_BUMP
@@ -228,6 +228,7 @@
     OccLists<Lit, vec<Watcher>, WatcherDeleted>
                         watches_bin,      // Watches for binary clauses only.
                         watches;          // 'watches[lit]' is a list of constraints watching 'lit' (will go there if literal becomes true).
+    vec<bool>           assump;           // Declares if a variable is an assumption variable or not.
     vec<lbool>          assigns;          // The current assignments.
     vec<char>           polarity;         // The preferred polarity of each variable.
     vec<char>           decision;         // Declares if a variable is eligible for selection in the decision heuristic.
@@ -284,7 +285,7 @@
     CRef     propagate        ();                                                      // Perform unit propagation. Returns possibly conflicting clause.
     void     cancelUntil      (int level);                                             // Backtrack until a certain level.
     void     analyze          (CRef confl, vec<Lit>& out_learnt, int& out_btlevel, int& out_lbd);    // (bt = backtrack)
-    void     analyzeFinal     (Lit p, vec<Lit>& out_conflict);                         // COULD THIS BE IMPLEMENTED BY THE ORDINARIY "analyze" BY SOME REASONABLE GENERALIZATION?
+    void     analyzeFinal(CRef confl, vec<Lit>& out_conflict);
     bool     litRedundant     (Lit p, uint32_t abstract_levels);                       // (helper method for 'analyze()')
     lbool    search           (int& nof_conflicts);                                    // Search for a given number of conflicts.
     lbool    solve_           ();                                                      // Main solve method (assumptions given in 'assumptions').
diff -urB cominisatps-SAT2016/core/SolverTypes.h cominisatps/core/SolverTypes.h
--- cominisatps-SAT2016/core/SolverTypes.h	2016-05-02 03:22:00.000000000 +0200
+++ cominisatps/core/SolverTypes.h	2020-07-17 09:34:07.994400936 +0200
@@ -49,7 +49,7 @@
     int     x;
 
     // Use this as a constructor:
-    friend Lit mkLit(Var var, bool sign = false);
+    //friend Lit mkLit(Var var, bool sign = false);
 
     bool operator == (Lit p) const { return x == p.x; }
     bool operator != (Lit p) const { return x != p.x; }
@@ -57,7 +57,7 @@
 };
 
 
-inline  Lit  mkLit     (Var var, bool sign) { Lit p; p.x = var + var + (int)sign; return p; }
+inline  Lit  mkLit     (Var var, bool sign = false) { Lit p; p.x = var + var + (int)sign; return p; }
 inline  Lit  operator ~(Lit p)              { Lit q; q.x = p.x ^ 1; return q; }
 inline  Lit  operator ^(Lit p, bool b)      { Lit q; q.x = p.x ^ (unsigned int)b; return q; }
 inline  bool sign      (Lit p)              { return p.x & 1; }
diff -urB cominisatps-SAT2016/mtl/template.mk cominisatps/mtl/template.mk
--- cominisatps-SAT2016/mtl/template.mk	2016-05-02 03:22:00.000000000 +0200
+++ cominisatps/mtl/template.mk	2020-07-17 11:36:18.589345683 +0200
@@ -44,7 +44,7 @@
 %.o:			CFLAGS +=$(COPTIMIZE) -g -D DEBUG
 %.op:			CFLAGS +=$(COPTIMIZE) -pg -g -D NDEBUG
 %.od:			CFLAGS +=-O0 -g -D DEBUG
-%.or:			CFLAGS +=$(COPTIMIZE) -g -D NDEBUG
+%.or:			CFLAGS +=$(COPTIMIZE)  -D NDEBUG
 
 ## Link options
 $(EXEC):		LFLAGS += -g
@@ -95,7 +95,7 @@
 depend.mk: $(CSRCS) $(CHDRS)
 	@echo Making dependencies
 	@$(CXX) $(CFLAGS) -I$(MROOT) \
-	   $(CSRCS) -MM | sed 's|\(.*\):|$(PWD)/\1 $(PWD)/\1r $(PWD)/\1d $(PWD)/\1p:|' > depend.mk
+	   $(CSRCS) -MM | sed 's|\([^:]*\):|$(PWD)/\1 $(PWD)/\1r $(PWD)/\1d $(PWD)/\1p:|' > depend.mk
 	@for dir in $(DEPDIR); do \
 	      if [ -r $(MROOT)/$${dir}/depend.mk ]; then \
 		  echo Depends on: $${dir}; \
diff -urB cominisatps-SAT2016/simp/Main.cc cominisatps/simp/Main.cc
--- cominisatps-SAT2016/simp/Main.cc	2016-05-02 03:22:00.000000000 +0200
+++ cominisatps/simp/Main.cc	2020-07-17 09:34:07.994400936 +0200
@@ -24,7 +24,7 @@
 
 #include <signal.h>
 #include <zlib.h>
-#include <sys/resource.h>
+//#include <sys/resource.h>
 
 #include "utils/System.h"
 #include "utils/ParseUtils.h"
@@ -41,11 +41,11 @@
 {
     double cpu_time = cpuTime();
     double mem_used = memUsedPeak();
-    printf("c restarts              : %"PRIu64"\n", solver.starts);
-    printf("c conflicts             : %-12"PRIu64"   (%.0f /sec)\n", solver.conflicts   , solver.conflicts   /cpu_time);
-    printf("c decisions             : %-12"PRIu64"   (%4.2f %% random) (%.0f /sec)\n", solver.decisions, (float)solver.rnd_decisions*100 / (float)solver.decisions, solver.decisions   /cpu_time);
-    printf("c propagations          : %-12"PRIu64"   (%.0f /sec)\n", solver.propagations, solver.propagations/cpu_time);
-    printf("c conflict literals     : %-12"PRIu64"   (%4.2f %% deleted)\n", solver.tot_literals, (solver.max_literals - solver.tot_literals)*100 / (double)solver.max_literals);
+    printf("c restarts              : %" PRIu64"\n", solver.starts);
+    printf("c conflicts             : %-12" PRIu64"   (%.0f /sec)\n", solver.conflicts   , solver.conflicts   /cpu_time);
+    printf("c decisions             : %-12" PRIu64"   (%4.2f %% random) (%.0f /sec)\n", solver.decisions, (float)solver.rnd_decisions*100 / (float)solver.decisions, solver.decisions   /cpu_time);
+    printf("c propagations          : %-12" PRIu64"   (%.0f /sec)\n", solver.propagations, solver.propagations/cpu_time);
+    printf("c conflict literals     : %-12" PRIu64"   (%4.2f %% deleted)\n", solver.tot_literals, (solver.max_literals - solver.tot_literals)*100 / (double)solver.max_literals);
     if (mem_used != 0) printf("c Memory used           : %.2f MB\n", mem_used);
     printf("c CPU time              : %g s\n", cpu_time);
 }
@@ -111,30 +111,12 @@
         solver = &S;
         // Use signal handlers that forcibly quit until the solver will be able to respond to
         // interrupts:
-        signal(SIGINT, SIGINT_exit);
-        signal(SIGXCPU,SIGINT_exit);
+        sigTerm(SIGINT_exit);
+
+        // Try to set resource limits:
+        if (cpu_lim != INT32_MAX) limitTime(cpu_lim);
+        if (mem_lim != INT32_MAX) limitMemory(mem_lim);
 
-        // Set limit on CPU-time:
-        if (cpu_lim != INT32_MAX){
-            rlimit rl;
-            getrlimit(RLIMIT_CPU, &rl);
-            if (rl.rlim_max == RLIM_INFINITY || (rlim_t)cpu_lim < rl.rlim_max){
-                rl.rlim_cur = cpu_lim;
-                if (setrlimit(RLIMIT_CPU, &rl) == -1)
-                    printf("c WARNING! Could not set resource limit: CPU-time.\n");
-            } }
-
-        // Set limit on virtual memory:
-        if (mem_lim != INT32_MAX){
-            rlim_t new_mem_lim = (rlim_t)mem_lim * 1024*1024;
-            rlimit rl;
-            getrlimit(RLIMIT_AS, &rl);
-            if (rl.rlim_max == RLIM_INFINITY || new_mem_lim < rl.rlim_max){
-                rl.rlim_cur = new_mem_lim;
-                if (setrlimit(RLIMIT_AS, &rl) == -1)
-                    printf("c WARNING! Could not set resource limit: Virtual memory.\n");
-            } }
-        
         if (argc == 1)
             printf("c Reading from standard input... Use '--help' for help.\n");
 
@@ -160,8 +142,7 @@
 
         // Change to signal-handlers that will only notify the solver and allow it to terminate
         // voluntarily:
-        signal(SIGINT, SIGINT_interrupt);
-        signal(SIGXCPU,SIGINT_interrupt);
+        sigTerm(SIGINT_interrupt);
 
         S.parsing = false;
         S.eliminate(true);
diff -urB cominisatps-SAT2016/simp/SimpSolver.cc cominisatps/simp/SimpSolver.cc
--- cominisatps-SAT2016/simp/SimpSolver.cc	2016-05-02 03:22:00.000000000 +0200
+++ cominisatps/simp/SimpSolver.cc	2020-07-17 09:34:07.994400936 +0200
@@ -636,6 +636,7 @@
     int iter = 0;
     int n_cls, n_cls_init, n_vars;
 
+    printf("c Using COMiniSatPS SAT solver by Chanseok Oh (2016)\n");
     if (nVars() == 0) goto cleanup; // User disabling preprocessing.
 
     // Get an initial number of clauses (more accurately).
@@ -648,10 +649,10 @@
     n_cls  = nClauses();
     n_vars = nFreeVars();
 
-    printf("c Reduced to %d vars, %d cls (grow=%d)\n", n_vars, n_cls, grow);
+    if (verbosity >= 1) printf("c COMiniSatPS: Reduced to %d vars, %d cls (grow=%d)\n", n_vars, n_cls, grow);
 
     if ((double)n_cls / n_vars >= 10 || n_vars < 10000){
-        printf("c No iterative elimination performed. (vars=%d, c/v ratio=%.1f)\n", n_vars, (double)n_cls / n_vars);
+        if (verbosity >= 1) printf("c COMiniSatPS: No iterative elimination performed. (vars=%d, c/v ratio=%.1f)\n", n_vars, (double)n_cls / n_vars);
         goto cleanup; }
 
     grow = grow ? grow * 2 : 8;
@@ -678,12 +679,12 @@
         double cl_inc_rate  = (double)n_cls_now   / n_cls_last;
         double var_dec_rate = (double)n_vars_last / n_vars_now;
 
-        printf("c Reduced to %d vars, %d cls (grow=%d)\n", n_vars_now, n_cls_now, grow);
-        printf("c cl_inc_rate=%.3f, var_dec_rate=%.3f\n", cl_inc_rate, var_dec_rate);
+        if (verbosity >= 1) printf("c COMiniSatPS: Reduced to %d vars, %d cls (grow=%d)\n", n_vars_now, n_cls_now, grow);
+        if (verbosity >= 1) printf("c COMiniSatPS: cl_inc_rate=%.3f, var_dec_rate=%.3f\n", cl_inc_rate, var_dec_rate);
 
         if (n_cls_now > n_cls_init || cl_inc_rate > var_dec_rate) break;
     }
-    printf("c No. effective iterative eliminations: %d\n", iter);
+    if (verbosity >= 1) printf("c COMiniSatPS: No. effective iterative eliminations: %d\n", iter);
 
 cleanup:
     touched  .clear(true);
diff -urB cominisatps-SAT2016/utils/Options.h cominisatps/utils/Options.h
--- cominisatps-SAT2016/utils/Options.h	2016-05-02 03:22:00.000000000 +0200
+++ cominisatps/utils/Options.h	2020-07-17 09:34:07.994400936 +0200
@@ -282,15 +282,15 @@
         if (range.begin == INT64_MIN)
             fprintf(stderr, "imin");
         else
-            fprintf(stderr, "%4"PRIi64, range.begin);
+            fprintf(stderr, "%4" PRIi64, range.begin);
 
         fprintf(stderr, " .. ");
         if (range.end == INT64_MAX)
             fprintf(stderr, "imax");
         else
-            fprintf(stderr, "%4"PRIi64, range.end);
+            fprintf(stderr, "%4" PRIi64, range.end);
 
-        fprintf(stderr, "] (default: %"PRIi64")\n", value);
+        fprintf(stderr, "] (default: %" PRIi64")\n", value);
         if (verbose){
             fprintf(stderr, "\n        %s\n", description);
             fprintf(stderr, "\n");
diff -urB cominisatps-SAT2016/utils/System.cc cominisatps/utils/System.cc
--- cominisatps-SAT2016/utils/System.cc	2016-05-02 03:22:00.000000000 +0200
+++ cominisatps/utils/System.cc	2020-10-21 11:29:27.654831916 +0200
@@ -18,11 +18,13 @@
 OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 **************************************************************************************************/
 
+#include <signal.h>
+#include <stdio.h>
+
 #include "utils/System.h"
 
 #if defined(__linux__)
 
-#include <stdio.h>
 #include <stdlib.h>
 
 using namespace Minisat;
@@ -88,8 +90,73 @@
     malloc_statistics_t t;
     malloc_zone_statistics(NULL, &t);
     return (double)t.max_size_in_use / (1024*1024); }
+double MiniSat::memUsedPeak(void) { return memUsed(); }
 
 #else
 double Minisat::memUsed() { 
     return 0; }
+double Minisat::memUsedPeak() { 
+    return 0; }
+#endif
+
+#if !defined(_MSC_VER) && !defined(__MINGW32__)
+void Minisat::limitMemory(uint64_t max_mem_mb)
+{
+// FIXME: OpenBSD does not support RLIMIT_AS. Not sure how well RLIMIT_DATA works instead.
+#if defined(__OpenBSD__)
+#define RLIMIT_AS RLIMIT_DATA
+#endif
+
+    // Set limit on virtual memory:
+    if (max_mem_mb != 0){
+        rlim_t new_mem_lim = (rlim_t)max_mem_mb * 1024*1024;
+        rlimit rl;
+        getrlimit(RLIMIT_AS, &rl);
+        if (rl.rlim_max == RLIM_INFINITY || new_mem_lim < rl.rlim_max){
+            rl.rlim_cur = new_mem_lim;
+            if (setrlimit(RLIMIT_AS, &rl) == -1)
+                printf("WARNING! Could not set resource limit: Virtual memory.\n");
+        }
+    }
+
+#if defined(__OpenBSD__)
+#undef RLIMIT_AS
+#endif
+}
+#else
+void Minisat::limitMemory(uint64_t /*max_mem_mb*/)
+{
+    printf("WARNING! Memory limit not supported on this architecture.\n");
+}
 #endif
+
+
+#if !defined(_MSC_VER) && !defined(__MINGW32__)
+void Minisat::limitTime(uint32_t max_cpu_time)
+{
+    if (max_cpu_time != 0){
+        rlimit rl;
+        getrlimit(RLIMIT_CPU, &rl);
+        if (rl.rlim_max == RLIM_INFINITY || (rlim_t)max_cpu_time < rl.rlim_max){
+            rl.rlim_cur = max_cpu_time;
+            if (setrlimit(RLIMIT_CPU, &rl) == -1)
+                printf("WARNING! Could not set resource limit: CPU-time.\n");
+        }
+    }
+}
+#else
+void Minisat::limitTime(uint32_t /*max_cpu_time*/)
+{
+    printf("WARNING! CPU-time limit not supported on this architecture.\n");
+}
+#endif
+
+
+void Minisat::sigTerm(void handler(int))
+{
+    signal(SIGINT, handler);
+    signal(SIGTERM,handler);
+#ifdef SIGXCPU
+    signal(SIGXCPU,handler);
+#endif
+}
diff -urB cominisatps-SAT2016/utils/System.h cominisatps/utils/System.h
--- cominisatps-SAT2016/utils/System.h	2016-05-02 03:22:00.000000000 +0200
+++ cominisatps/utils/System.h	2020-07-17 09:34:07.994400936 +0200
@@ -34,6 +34,11 @@
 static inline double cpuTime(void); // CPU-time in seconds.
 extern double memUsed();            // Memory in mega bytes (returns 0 for unsupported architectures).
 extern double memUsedPeak();        // Peak-memory in mega bytes (returns 0 for unsupported architectures).
+extern void   limitMemory(uint64_t max_mem_mb); // Set a limit on total memory usage. The exact
+                                                // semantics varies depending on architecture.
+extern void   limitTime(uint32_t max_cpu_time); // Set a limit on maximum CPU time. The exact
+                                                // semantics varies depending on architecture.
+extern void   sigTerm(void handler(int));      // Set up handling of available termination signals.
 
 }
 
